infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -452,13 +467,43 @@ public String evaluateLegacyNamespaceName(NamespaceResolutionContext resolutionC
    */
   public String evaluateNamespaceName(NamespaceResolutionContext resolutionCtx)
       throws InfrastructureException {
-    String namespace = evalPlaceholders(defaultNamespaceName, resolutionCtx);
+    String namespace;
+    Optional<Pair<String, String>> namespaceOptional =
```

Comment:
This is a minor nitpick. More of an observation than change request.

I like optionals for their composability. E.g. mapping them to other types, etc. We're not doing that here - we're just using it as a wrapper for a nullable value. Also, optionals are IMHO more usually used in the public API of types than in just private implementation details.

So what about just handling the `Pair<String, String>` as a nullable value and saving 1 allocation?

Reply:
i know at least one guy who will be unhappy with nulls :)

Reply:
I don't think saving 1 allocation will improve anything here. However, the code here looks a bit overcomplicated. I would maybe move the `allowUserDefinedNamespaces` check to `getPreferencesNamespaceName` method, and rename `namespaceOptional` to `storedNamespace` ?